test/e2e/framework/cluster.go

```diff
@@ -0,0 +1,352 @@
+// Copyright (C) 2024 ScyllaDB
+
+package framework
+
+import (
+	"context"
+	"crypto/sha512"
+	"encoding/base64"
+	"fmt"
+	"os"
+	"path"
+	"strconv"
+	"time"
+
+	g "github.com/onsi/ginkgo/v2"
+	o "github.com/onsi/gomega"
+	scyllaclientset "github.com/scylladb/scylla-operator/pkg/client/scylla/clientset/versioned"
+	"github.com/scylladb/scylla-operator/pkg/controllerhelpers"
+	corev1 "k8s.io/api/core/v1"
+	rbacv1 "k8s.io/api/rbac/v1"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/util/wait"
+	"k8s.io/apiserver/pkg/storage/names"
+	"k8s.io/client-go/discovery"
+	cacheddiscovery "k8s.io/client-go/discovery/cached/memory"
+	"k8s.io/client-go/dynamic"
+	"k8s.io/client-go/kubernetes"
+	restclient "k8s.io/client-go/rest"
+	"k8s.io/klog/v2"
+)
+
+type Cluster interface {
+	Namespace() string
+	Username() string
+	FieldManager() string
+	ClientConfig() *restclient.Config
+	AdminClientConfig() *restclient.Config
+	DiscoveryClient() *discovery.DiscoveryClient
+	DynamicClient() dynamic.Interface
+	DynamicAdminClient() dynamic.Interface
+	KubeClient() *kubernetes.Clientset
+	KubeAdminClient() *kubernetes.Clientset
+	ScyllaClient() *scyllaclientset.Clientset
+	ScyllaAdminClient() *scyllaclientset.Clientset
+	CommonLabels() map[string]string
+	SetUp(ctx context.Context)
```

Comment:
CreateNamespace might be more useful. You may also need more then one.

Reply:
This isn't just creating the namespace but also setting up a service account, rbac and creating a restricted client. So the following methods are all depending on it:
```
	ClientConfig() *restclient.Config
	DiscoveryClient() *discovery.DiscoveryClient
	DynamicClient() dynamic.Interface
	KubeClient() *kubernetes.Clientset
	ScyllaClient() *scyllaclientset.Clientset
```
So I don't know what exactly you have in mind. If you want me to implement it exactly the way you'd do it, please describe it in more detail, I don't want to do n more iterations on this PR playing a guessing game.

Reply:
Discussed offline to sync - we'll most likely need something like a Namespace abstraction, but we have to maintain backwards compatibility with Framework to not have to change the existing tests. I think we could move the above methods away from Cluster interface to a Namespace interface, and have Framework implement this interface as well.

Reply:
I took a stab at it at 
https://github.com/tnozicka/scylla-operator/commits/kacper-multi-cluster-framework/ 
feel free to pick that as a starting point and make changes to make it work if needed - I only took it to the point where it compiled to see the abstraction make sense

Reply:
@tnozicka I don't get the `Interface` suffix in interface types and changing the `cluster` struct to public `Cluster`, can you explain?

Reply:
also why would `Cluster` interface still have the insecure Namespace method
```
func (c *Cluster) Namespace() string {
	return c.namespace.Name
}
```
if only the default one should have the default namespace?
You also added the `BeforeEach` and `AfterEach` funcs to Cluster implementation, which would imply they should be run, but then I only see them run for the default cluster - imo that's counter-intuitive. If you only want them run for the default cluster, they shouldn't be `Cluster`'s methods, but `Framework`'s.

Reply:
> @tnozicka I don't get the Interface suffix in interface types and changing the cluster struct to public Cluster, can you explain?

I made them public, so I needed to distinguish them. The `client` was also too often colliding and shadowing with variable names.

> also why would Cluster interface still have the insecure Namespace method

It goes into the frawework.FieldManager for a similar point as bellow

> if only the default one should have the default namespace?
You also added the BeforeEach and AfterEach funcs to Cluster implementation, which would imply they should be run, but then I only see them run for the default cluster - imo that's counter-intuitive. If you only want them run for the default cluster, they shouldn't be Cluster's methods, but Framework's.

Some of this could be on the Framework but at that point you loose the `ClientInterface` on the `Cluster`, so it's a choice to make.

Happy to chat about this tomorrow and see your point of view.

Reply:
> > also why would Cluster interface still have the insecure Namespace method
> 
> It goes into the frawework.FieldManager for a similar point as bellow

ok, but why make it insecure? I just don't get the point of these two methods
```
func (c *Cluster) DefaultNamespaceIfAny() string {
	o.Expect(c.namespace).NotTo(o.BeNil())
	return c.namespace.Name
}

func (c *Cluster) Namespace() string {
	return c.namespace.Name
}
```

if there's a reason for keeping both I'd expect something like
```
func (c *Cluster) DefaultNamespaceIfAny() (string, bool) {
	if c.Namespace == nil {
	  return "", false
	}
	
	return c.Namespace.Name, true
}

func (c *Cluster) Namespace() string {
	o.Expect(c.namespace).NotTo(o.BeNil())
	return c.Namespace.Name
}
```

Reply:
@tnozicka I applied the changes. Changed some things here and there but the overall approach stayed the same. Notable changes that I remember wrt your suggestion:
- cluster doesn't have a Namespace method - didn't make sense if only the default cluster has a default namespace and it's enforced by framework
- cluster only has AdminClient, not Client - same point as above
- moved fieldManager away from cluster - same point as above
- cluster doesn't have beforeEach/afterEach method - that only made sense for the default cluster, which is only a concept in the context of framework, so imo framework should be doing that work

You also left a FIXME in namespace deletion func - was that referring to movinf artifacts dir creation away from it? Because imo it's ok to do it there, even if you end up doing it multiple times unnecessarily.

Reply:
> ok, but why make it insecure? I just don't get the point of these two methods

It's all to have the `FullClientInterface` on the `Cluster`.

I found it a bit confusing for the `Framework` to implement `FullClientInterface` and when you do `Framework.DefaultCluster` you reduce it to `AdminClientInterface`

essentially all the rest is a consequence of this choice

Reply:
ok, let's try it with just the admin client, it's a valid option